File: sdks/rust/src/permissions.rs
use serde::{Deserialize, Serialize};

/// Permission types in IORA
#[derive(Debug, Clone, Copy, PartialEq, Eq, Hash, Serialize, Deserialize)]
pub enum Permission {
    // Entity permissions
    ReadEntities,
    ControlEntities,
    CreateEntities,
    DeleteEntities,

    // Storage permissions
    StorageRead,
    StorageWrite,
    StorageDelete,

    // Network permissions
    NetworkAccess,
    NetworkOutbound,
    NetworkInbound,
    NetworkScan,
    NetworkLocalAccess,

    // System permissions
    SystemInfo,
    SystemControl,
    SystemRestart,

    // Database permissions
    DatabaseRead,
    DatabaseWrite,
    DatabaseCreate,
    DatabaseDelete,

    // API and Widget permissions
    RegisterApi,
    CallApi,
    RegisterWidget,
    ControlWidget,

    // Notification permissions
    SendNotifications,
    ReadNotifications,

    // Media permissions
    CameraAccess,
    MicrophoneAccess,
    MediaAccess,

    // Advanced permissions
    PluginManager,
    InstallPlugins,
    FileSystemRead,
    FileSystemWrite,
    FileSystemExecute,
    LocationAccess,
    Automations,

    // File sharing permissions (iora-share)
    FileShareRead,
    FileShareWrite,
    FileShareDelete,
    FileShareManage,

    // Advanced app-only permissions
    BackupAccess,
    LogAccess,
    UserManagement,
    SecuritySettings,
    NetworkMonitoring,
    ProcessControl,
}

impl Permission {
    /// Get the risk level of this permission
    pub fn risk_level(&self) -> RiskLevel {
        match self {
            Permission::ReadEntities | Permission::StorageRead | Permission::DatabaseRead
            | Permission::SystemInfo | Permission::ReadNotifications => RiskLevel::Low,

            Permission::ControlEntities | Permission::StorageWrite | Permission::DatabaseWrite
            | Permission::CallApi | Permission::SendNotifications | Permission::MediaAccess => RiskLevel::Medium,

            Permission::CreateEntities | Permission::DeleteEntities | Permission::NetworkAccess
            | Permission::NetworkOutbound | Permission::RegisterApi | Permission::RegisterWidget
            | Permission::CameraAccess | Permission::FileSystemRead | Permission::FileShareRead
            | Permission::LogAccess | Permission::BackupAccess => RiskLevel::High,

            Permission::SystemControl | Permission::SystemRestart | Permission::PluginManager
            | Permission::InstallPlugins | Permission::FileSystemWrite | Permission::FileSystemExecute
            | Permission::NetworkScan | Permission::NetworkLocalAccess | Permission::FileShareWrite
            | Permission::FileShareDelete | Permission::FileShareManage | Permission::UserManagement
            | Permission::SecuritySettings | Permission::NetworkMonitoring | Permission::ProcessControl => RiskLevel::Critical,

            _ => RiskLevel::Medium,
        }
    }

    /// Get the permission category (App-only, Plugin-allowed, requires consent)
    pub fn category(&self) -> PermissionCategory {
        match self {
            // Plugin-allowed permissions (basic functionality)
            Permission::ReadEntities | Permission::ControlEntities | Permission::StorageRead
            | Permission::StorageWrite | Permission::NetworkOutbound | Permission::SystemInfo
            | Permission::CallApi | Permission::SendNotifications | Permission::ReadNotifications
            | Permission::MediaAccess => PermissionCategory::PluginAllowed,

            // App-only permissions (advanced functionality, auto-approved)
            Permission::CreateEntities | Permission::DeleteEntities | Permission::StorageDelete
            | Permission::NetworkAccess | Permission::DatabaseRead | Permission::DatabaseWrite
            | Permission::RegisterApi | Permission::RegisterWidget | Permission::ControlWidget
            | Permission::Automations | Permission::FileShareRead => PermissionCategory::AppOnly,

            // App-only with user consent (sensitive operations)
            Permission::SystemControl | Permission::SystemRestart | Permission::PluginManager
            | Permission::InstallPlugins | Permission::FileSystemRead | Permission::FileSystemWrite
            | Permission::FileSystemExecute | Permission::NetworkScan | Permission::NetworkLocalAccess
            | Permission::CameraAccess | Permission::MicrophoneAccess | Permission::LocationAccess
            | Permission::DatabaseCreate | Permission::DatabaseDelete | Permission::FileShareWrite
            | Permission::FileShareDelete | Permission::FileShareManage | Permission::BackupAccess
            | Permission::LogAccess | Permission::UserManagement | Permission::SecuritySettings
            | Permission::NetworkMonitoring | Permission::ProcessControl | Permission::NetworkInbound => PermissionCategory::AppOnlyWithConsent,
        }
    }

    /// Check if this permission is available to plugins
    pub fn is_plugin_allowed(&self) -> bool {
        matches!(self.category(), PermissionCategory::PluginAllowed)
    }

    /// Check if this permission requires explicit user consent
    pub fn requires_user_consent(&self) -> bool {
        matches!(self.category(), PermissionCategory::AppOnlyWithConsent)
    }

    /// Get a human-readable description of this permission
    pub fn description(&self) -> &'static str {
        match self {
            Permission::ReadEntities => "Read entity states and attributes",
            Permission::ControlEntities => "Control smart home devices",
            Permission::CreateEntities => "Create new entities",
            Permission::DeleteEntities => "Delete entities",
            Permission::StorageRead => "Read from storage",
            Permission::StorageWrite => "Write to storage",
            Permission::StorageDelete => "Delete from storage",
            Permission::NetworkAccess => "Access external networks",
            Permission::NetworkOutbound => "Make outbound network requests",
            Permission::NetworkInbound => "Accept inbound network connections",
            Permission::NetworkScan => "Scan local network",
            Permission::NetworkLocalAccess => "Access local network devices",
            Permission::SystemInfo => "Read system information",
            Permission::SystemControl => "Control system settings",
            Permission::SystemRestart => "Restart IORA system",
            Permission::DatabaseRead => "Read from database",
            Permission::DatabaseWrite => "Write to database",
            Permission::DatabaseCreate => "Create database tables",
            Permission::DatabaseDelete => "Delete from database",
            Permission::RegisterApi => "Register API endpoints",
            Permission::CallApi => "Call API endpoints",
            Permission::RegisterWidget => "Register dashboard widgets",
            Permission::ControlWidget => "Control widgets",
            Permission::SendNotifications => "Send notifications",
            Permission::ReadNotifications => "Read notifications",
            Permission::CameraAccess => "Access camera",
            Permission::MicrophoneAccess => "Access microphone",
            Permission::MediaAccess => "Access media files",
            Permission::PluginManager => "Manage plugins",
            Permission::InstallPlugins => "Install/uninstall plugins",
            Permission::FileSystemRead => "Read files",
            Permission::FileSystemWrite => "Write files",
            Permission::FileSystemExecute => "Execute files",
            Permission::LocationAccess => "Access location data",
            Permission::Automations => "Create and manage automations",
            Permission::FileShareRead => "Read shared files via iora-share",
            Permission::FileShareWrite => "Upload files to iora-share",
            Permission::FileShareDelete => "Delete files from iora-share",
            Permission::FileShareManage => "Manage file sharing settings and permissions",
            Permission::BackupAccess => "Access system backups",
            Permission::LogAccess => "Access system and app logs",
            Permission::UserManagement => "Manage IORA users and accounts",
            Permission::SecuritySettings => "Modify security settings",
            Permission::NetworkMonitoring => "Monitor network traffic and connections",
            Permission::ProcessControl => "Control system processes and services",
        }
    }

    /// Get full metadata for this permission
    pub fn metadata(&self) -> PermissionMetadata {
        PermissionMetadata {
            permission: *self,
            category: self.category(),
            risk_level: self.risk_level(),
            description: self.description(),
        }
    }
}

#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize)]
pub enum RiskLevel {
    Low,
    Medium,
    High,
    Critical,
}

/// Permission category defining access level and approval requirements
#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize)]
pub enum PermissionCategory {
    /// Available to both Apps and Plugins, auto-approved based on manifest
    PluginAllowed,
    /// Only available to Apps, auto-approved based on manifest
    AppOnly,
    /// Only available to Apps, requires explicit user consent per installation
    AppOnlyWithConsent,
}

/// Permission metadata
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct PermissionMetadata {
    pub permission: Permission,
    pub category: PermissionCategory,
    pub risk_level: RiskLevel,
    pub description: &'static str,
}

File: sdks/php/src/Permissions.php
<?php

namespace rumahl;

/**
 * rumahl Permission types
 */
enum Permission: string
{
    // Entity permissions
    case READ_ENTITIES = 'ReadEntities';
    case CONTROL_ENTITIES = 'ControlEntities';
    case CREATE_ENTITIES = 'CreateEntities';
    case DELETE_ENTITIES = 'DeleteEntities';

    // Storage permissions
    case STORAGE_READ = 'StorageRead';
    case STORAGE_WRITE = 'StorageWrite';
    case STORAGE_DELETE = 'StorageDelete';

    // Network permissions
    case NETWORK_ACCESS = 'NetworkAccess';
    case NETWORK_OUTBOUND = 'NetworkOutbound';
    case NETWORK_INBOUND = 'NetworkInbound';
    case NETWORK_SCAN = 'NetworkScan';
    case NETWORK_LOCAL_ACCESS = 'NetworkLocalAccess';

    // System permissions
    case SYSTEM_INFO = 'SystemInfo';
    case SYSTEM_CONTROL = 'SystemControl';
    case SYSTEM_RESTART = 'SystemRestart';

    // Database permissions
    case DATABASE_READ = 'DatabaseRead';
    case DATABASE_WRITE = 'DatabaseWrite';
    case DATABASE_CREATE = 'DatabaseCreate';
    case DATABASE_DELETE = 'DatabaseDelete';

    // API and Widget permissions
    case REGISTER_API = 'RegisterApi';
    case CALL_API = 'CallApi';
    case REGISTER_WIDGET = 'RegisterWidget';
    case CONTROL_WIDGET = 'ControlWidget';

    // Notification permissions
    case SEND_NOTIFICATIONS = 'SendNotifications';
    case READ_NOTIFICATIONS = 'ReadNotifications';

    // Media permissions
    case CAMERA_ACCESS = 'CameraAccess';
    case MICROPHONE_ACCESS = 'MicrophoneAccess';
    case MEDIA_ACCESS = 'MediaAccess';

    // Plugin and filesystem permissions
    case PLUGIN_MANAGER = 'PluginManager';
    case INSTALL_PLUGINS = 'InstallPlugins';
    case FILE_SYSTEM_READ = 'FileSystemRead';
    case FILE_SYSTEM_WRITE = 'FileSystemWrite';
    case FILE_SYSTEM_EXECUTE = 'FileSystemExecute';

    // Other permissions
    case LOCATION_ACCESS = 'LocationAccess';
    case AUTOMATIONS = 'Automations';
}

/**
 * Risk level for permissions
 */
enum RiskLevel: string
{
    case LOW = 'Low';
    case MEDIUM = 'Medium';
    case HIGH = 'High';
    case CRITICAL = 'Critical';
}

/**
 * Get risk level for a permission
 */
function getPermissionRiskLevel(Permission $permission): RiskLevel
{
    $lowRisk = [
        Permission::READ_ENTITIES,
        Permission::STORAGE_READ,
        Permission::DATABASE_READ,
        Permission::SYSTEM_INFO,
        Permission::READ_NOTIFICATIONS,
    ];

    $criticalRisk = [
        Permission::SYSTEM_CONTROL,
        Permission::SYSTEM_RESTART,
        Permission::PLUGIN_MANAGER,
        Permission::INSTALL_PLUGINS,
        Permission::FILE_SYSTEM_WRITE,
        Permission::FILE_SYSTEM_EXECUTE,
        Permission::NETWORK_SCAN,
        Permission::NETWORK_LOCAL_ACCESS,
    ];

    $highRisk = [
        Permission::CREATE_ENTITIES,
        Permission::DELETE_ENTITIES,
        Permission::NETWORK_ACCESS,
        Permission::NETWORK_OUTBOUND,
        Permission::REGISTER_API,
        Permission::REGISTER_WIDGET,
        Permission::CAMERA_ACCESS,
        Permission::FILE_SYSTEM_READ,
    ];

    if (in_array($permission, $lowRisk, true)) {
        return RiskLevel::LOW;
    }
    if (in_array($permission, $criticalRisk, true)) {
        return RiskLevel::CRITICAL;
    }
    if (in_array($permission, $highRisk, true)) {
        return RiskLevel::HIGH;
    }

    return RiskLevel::MEDIUM;
}

/**
 * Get description for a permission
 */
function getPermissionDescription(Permission $permission): string
{
    return match ($permission) {
        Permission::READ_ENTITIES => 'Read entity states and attributes',
        Permission::CONTROL_ENTITIES => 'Control smart home devices',
        Permission::CREATE_ENTITIES => 'Create new entities',
        Permission::DELETE_ENTITIES => 'Delete entities',
        Permission::STORAGE_READ => 'Read from storage',
        Permission::STORAGE_WRITE => 'Write to storage',
        Permission::STORAGE_DELETE => 'Delete from storage',
        Permission::NETWORK_ACCESS => 'Access external networks',
        Permission::NETWORK_OUTBOUND => 'Make outbound network requests',
        Permission::NETWORK_INBOUND => 'Accept inbound network connections',
        Permission::NETWORK_SCAN => 'Scan local network',
        Permission::NETWORK_LOCAL_ACCESS => 'Access local network devices',
        Permission::SYSTEM_INFO => 'Read system information',
        Permission::SYSTEM_CONTROL => 'Control system settings',
        Permission::SYSTEM_RESTART => 'Restart rumahl system',
        Permission::DATABASE_READ => 'Read from database',
        Permission::DATABASE_WRITE => 'Write to database',
        Permission::DATABASE_CREATE => 'Create database tables',
        Permission::DATABASE_DELETE => 'Delete from database',
        Permission::REGISTER_API => 'Register API endpoints',
        Permission::CALL_API => 'Call API endpoints',
        Permission::REGISTER_WIDGET => 'Register dashboard widgets',
        Permission::CONTROL_WIDGET => 'Control widgets',
        Permission::SEND_NOTIFICATIONS => 'Send notifications',
        Permission::READ_NOTIFICATIONS => 'Read notifications',
        Permission::CAMERA_ACCESS => 'Access camera',
        Permission::MICROPHONE_ACCESS => 'Access microphone',
        Permission::MEDIA_ACCESS => 'Access media files',
        Permission::PLUGIN_MANAGER => 'Manage plugins',
        Permission::INSTALL_PLUGINS => 'Install/uninstall plugins',
        Permission::FILE_SYSTEM_READ => 'Read files',
        Permission::FILE_SYSTEM_WRITE => 'Write files',
        Permission::FILE_SYSTEM_EXECUTE => 'Execute files',
        Permission::LOCATION_ACCESS => 'Access location data',
        Permission::AUTOMATIONS => 'Create and manage automations',
        default => 'Unknown permission',
    };
}
